### PR TITLE
Add `no_warning_propagation` flag to `Vector.build` and `Builder.new`

### LINKED ISSUE #10046: Vector benchmarks regression
Screenshot from 2024-05-23 11-34-13
Screenshot from 2024-05-23 11-32-32

### PR DESCRIPTION
**Pull request by GregoryTravis**
Fixes #10046.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java [hunk 1]
```diff
@@ -38,22 +38,23 @@ public void initializeBenchmark(BenchmarkParams params) throws Exception {
...
         avg arr =
             sum acc i = if i == arr.length then acc else
                 @Tail_Call sum (acc + arr.at i) i+1
             (sum 0 0) / arr.length
 
         fibarr size modulo =
-            b = Builder.new size
-            b.append 1
-            b.append 1
+            Vector.build initial_capacity=size no_warning_propagation=True b->
```
Comment (GregoryTravis) [inline engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java]: This change does not affect the speed in any way; it's merely updating the code to the preferred style.

**Review by 4e6 — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso [hunk 2]
```diff
@@ -171,6 +171,8 @@ type Vector a
        - function: a function taking a `Builder` and adding elements to it.
        - initial_capacity: Initial capacity of the `Builder`; this is for
          pre-allocation and does not affect the size of the resulting `Vector`.
+       - no_warning_propagation: If True, do not gather warnings from values
+         added to the builder.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: Wouldn't it be easier if the paramater was `warning_propagation_enabled` or `propagate_warnings`? Negation in boolean flags usually makes it more complicated as it requires a double negation - if `no_X` is `False` then `X` is enabled. It is much clearer if it is just `X` is true or false.

File: engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java (hunk 1 above)
Comment (radeusgd) [reply engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java]: Shall we have 2 variants of this benchmark running? One with and one without benchmark propagation? Then we could compare and see how big the overhead of propagation is.

Comment (radeusgd) [reply engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java]: We would also be able to track how both 'modes' behave over time.

**Review by radeusgd — APPROVED** [review]
Looks good, but I think we should be measuring both _with_ and _without_ warnings propagation, to see the whole picture. Also I think it would be good to schedule a benchmark run before the PR is merged, so that we can see what the results are.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso (hunk 2 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: Changed to `propagate_warnings`.

File: engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java (hunk 1 above)
Comment (GregoryTravis) [reply engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java]: The purpose of this benchmark is mainly to measure the `avg` operation, so I added a separate benchmark at `test/Benchmarks/src/Vector/Build.enso` for just building vectors with and without warning propagation.

Comment (radeusgd) [reply engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java]: Ah right. Perfect. Can you post results of that benchmark?

File: test/Benchmarks/src/Vector/Build.enso [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            group_builder.specify "Builder_New_Append_One"+suffix <|
+                builder = Builder.new propagate_warnings=propagate_warnings
+                0.up_to num_values . map i->
+                    builder.append i
+                builder.to_vector
+            group_builder.specify "Builder_New_Append_Batch"+suffix <|
+                builder = Builder.new propagate_warnings=propagate_warnings
+                0.up_to num_batches . map _->
+                    builder.append_vector_range data.batch
+                builder.to_vector
```
Comment (radeusgd) [inline test/Benchmarks/src/Vector/Build.enso:49]: Great!

**Review by radeusgd — APPROVED** [review]

**Comment by GregoryTravis** [comment]
@JaroslavTulach 
I'm still seeing a difference in the benchmarks, from a change to `fibarr`. [This branch](https://github.com/enso-org/enso/tree/wip/gmt/10046-flag) adds the `propagate_warnings` flag, and uses it to speed up `VectorBenchmarks.averageOverSlice`. [This branch](https://github.com/enso-org/enso/tree/wip/gmt/10046-flag-reverted) reverts *only that change*.
Benchmark results:
```
wip/gmt/10046-flag:[info] VectorBenchmarks.averageOverSlice  avgt    5  0.001 ±  0.001  ms/op
wip/gmt/10046-flag-reverted:[info] VectorBenchmarks.averageOverSlice  avgt    5  0.006 ± 0.001  ms/op
```
Since this change is to code that is only used in setup, I wonder why it is affecting the benchmark results.

**Comment by JaroslavTulach** [comment]
The difference is in `propagate_warnings=False`. If it is there, the code runs fast. If it is not, it is **eight times slower**.

> Since this change is to code that is only used in setup, I wonder why it is affecting the benchmark results.

The only explanation is that the _setup creates different vector_ in one case and different one in the other case.

**Comment by JaroslavTulach** [comment]
> The only explanation is that the _setup creates different vector_ in one case and different one in the other case.

### Fast Case

This is what is needed to compute `arr.length` in the faster case:

1ms case

Please note that we have `long[]` wrapped in `Long.storage` wrapped inside of `Generic.storage` which is wrapped by `ArraySlice.storage` which is again wrapped by `Generic.storage`. It is interesting that even such level of indirection can be fast.

### Slow Case

This is what is needed to compute `arr.length` in the slower case:

8ms case

The slow case is more complicated. It contains **two if statements** that divide the code into **four branches**. I am not sure what is causing that, but clearly the `if` instructions have non-trivial cost and seem to be the reason why we are are seeing the slowdown.

**Comment by JaroslavTulach** [comment]
With `propagate_warnings=False` there is no call to [constructor of Array](https://github.com/enso-org/enso/blob/9027596b1b92b398bd8a4d2c5ccb358aa2b7b422/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Array.java#L42). Without that argument, there are many calls.

- #10143

The above PR has a potential to speed the slow case by 30% making it just **five times** slower than the propagate_warnings=False` case.

**Comment by JaroslavTulach** [comment]
> The slow case is more complicated. It contains **two if statements** that divide the code into **four branches**. I am not sure what is causing that

Debugging (as [described at this comment](https://github.com/enso-org/enso/pull/10143#issuecomment-2141242871)) shows the problem is in [Array_Like_Helpers.length](https://github.com/enso-org/enso/blob/41c82e06b35e453ed5d09c809cf7dcbd55f39c7f/distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Array_Like_Helpers.enso#L39). The `Builtin_Method "Array_Like_Helpers.length"` implementation node [LengthVectorNode](https://github.com/enso-org/enso/blob/41c82e06b35e453ed5d09c809cf7dcbd55f39c7f/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/LengthVectorNode.java#L11) **isn't split** but is **shared** between the benchmark's `avg` code and the code that _computes warnings_ that is invoked unless disabled by `propagate_warnings=False`.

Such a **sharing** of [LengthVectorNode](https://github.com/enso-org/enso/blob/41c82e06b35e453ed5d09c809cf7dcbd55f39c7f/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/LengthVectorNode.java#L11) is **unfortunate**. The PR ...

- #10143

mitigates the problem by using better `Vector` implementations (so we don't need `if` statements to switch between many different ones). The benchmark shall be fast. But it _the splitting problem_ isn't solved by the #10143 PR.

**Comment by JaroslavTulach** [comment]
> Such a **sharing** of [LengthVectorNode](https://github.com/enso-org/enso/blob/41c82e06b35e453ed5d09c809cf7dcbd55f39c7f/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/LengthVectorNode.java#L11) is **unfortunate**.

@Akirathan suggested to `@ReportPolymorphism`, I tried, but it doesn't seem to have any impact on the **splitting**:
```diff
diff --git engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java
index f845b53fef..e3124103ce 100644
--- engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java
+++ engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java
@@ -36,6 +36,12 @@ public abstract class BuiltinRootNode extends RootNode {
     this.typeName = typeName;
   }

+  @Override
+  public boolean isCloningAllowed() {
+    System.err.println("isClonning allowed: " + getClass().getName());
+    return true;
+  }
+
   /**
    * Executes this node's logic, returning a pair of return value and the new state.
    *
diff --git engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeAtNode.java engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeAtNode.java
index 5c097d24d4..30924b3650 100644
--- engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeAtNode.java
+++ engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeAtNode.java
@@ -2,6 +2,7 @@ package org.enso.interpreter.runtime.data.vector;

 import com.oracle.truffle.api.dsl.Cached;
 import com.oracle.truffle.api.dsl.NeverDefault;
+import com.oracle.truffle.api.dsl.ReportPolymorphism;
 import com.oracle.truffle.api.dsl.Specialization;
 import com.oracle.truffle.api.exception.AbstractTruffleException;
 import com.oracle.truffle.api.interop.InteropLibrary;
@@ -12,6 +13,7 @@ import com.oracle.truffle.api.nodes.Node;
 import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
 import org.enso.interpreter.runtime.error.WarningsLibrary;

+@ReportPolymorphism
 public abstract class ArrayLikeAtNode extends Node {
   public abstract Object executeAt(Object arrayLike, long index) throws InvalidArrayIndexException;

diff --git engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeLengthNode.java engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeLengthNode.java
index 38a5ce9daa..3437890e38 100644
--- engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeLengthNode.java
+++ engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeLengthNode.java
@@ -3,12 +3,14 @@ package org.enso.interpreter.runtime.data.vector;
 import com.oracle.truffle.api.dsl.Cached.Exclusive;
 import com.oracle.truffle.api.dsl.Fallback;
 import com.oracle.truffle.api.dsl.NeverDefault;
+import com.oracle.truffle.api.dsl.ReportPolymorphism;
 import com.oracle.truffle.api.dsl.Specialization;
 import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.UnsupportedMessageException;
 import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.nodes.Node;

+@ReportPolymorphism
 public abstract class ArrayLikeLengthNode extends Node {
   public abstract long executeLength(Object arrayLike);

```

File: engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java (hunk 1 above)
Comment (GregoryTravis) [reply engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java]: Here are two runs on this branch, and one run on this branch with @JaroslavTulach's #10143 included.

I think the overhead from propagating warnings is a result of the fact that it requires traversing collections and will have to be addressed by changes to warning propagation.

```
==> vb-flag-0-out <==
[info] Vector_Builder.Builder_New_Append_Batch_With_Warning_Propagation      avgt   10   30.661 ± 30.958  ms/op
[info] Vector_Builder.Builder_New_Append_Batch_Without_Warning_Propagation   avgt   10   18.598 ±  5.742  ms/op
[info] Vector_Builder.Builder_New_Append_One_With_Warning_Propagation        avgt   10  124.009 ± 51.695  ms/op
[info] Vector_Builder.Builder_New_Append_One_Without_Warning_Propagation     avgt   10   98.917 ± 37.554  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_With_Warning_Propagation     avgt   10   18.311 ±  4.139  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_Without_Warning_Propagation  avgt   10   12.310 ±  1.056  ms/op
[info] Vector_Builder.Vector_Build_Append_One_With_Warning_Propagation       avgt   10  241.134 ± 68.479  ms/op
[info] Vector_Builder.Vector_Build_Append_One_Without_Warning_Propagation    avgt   10   86.485 ± 53.192  ms/op
[info] Benchmark results reported into /Users/gmt/dev/enso/enso/std-bits/benchmarks/./bench-report.xml
[success] Total time: 609 s (10:09), completed May 31, 2024, 3:50:07 PM

==> vb-flag-1-out <==
[info] Vector_Builder.Builder_New_Append_Batch_With_Warning_Propagation      avgt   10   28.612 ±  5.543  ms/op
[info] Vector_Builder.Builder_New_Append_Batch_Without_Warning_Propagation   avgt   10   12.082 ±  2.142  ms/op
[info] Vector_Builder.Builder_New_Append_One_With_Warning_Propagation        avgt   10  116.839 ± 42.429  ms/op
[info] Vector_Builder.Builder_New_Append_One_Without_Warning_Propagation     avgt   10  112.696 ± 46.969  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_With_Warning_Propagation     avgt   10   17.174 ±  2.912  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_Without_Warning_Propagation  avgt   10   15.813 ±  1.365  ms/op
[info] Vector_Builder.Vector_Build_Append_One_With_Warning_Propagation       avgt   10  183.215 ±  9.186  ms/op
[info] Vector_Builder.Vector_Build_Append_One_Without_Warning_Propagation    avgt   10   91.330 ± 52.272  ms/op
[info] Benchmark results reported into /Users/gmt/dev/enso/enso/std-bits/benchmarks/./bench-report.xml
[success] Total time: 612 s (10:12), completed May 31, 2024, 4:19:41 PM

==> vb-flag-with-jt-out <==
[info] Vector_Builder.Builder_New_Append_Batch_With_Warning_Propagation      avgt   10    9.665 ±  0.994  ms/op
[info] Vector_Builder.Builder_New_Append_Batch_Without_Warning_Propagation   avgt   10   17.479 ±  4.903  ms/op
[info] Vector_Builder.Builder_New_Append_One_With_Warning_Propagation        avgt   10  116.431 ± 46.189  ms/op
[info] Vector_Builder.Builder_New_Append_One_Without_Warning_Propagation     avgt   10  100.541 ± 31.344  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_With_Warning_Propagation     avgt   10   26.978 ± 11.994  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_Without_Warning_Propagation  avgt   10   27.681 ± 13.834  ms/op
[info] Vector_Builder.Vector_Build_Append_One_With_Warning_Propagation       avgt   10  212.782 ± 17.657  ms/op
[info] Vector_Builder.Vector_Build_Append_One_Without_Warning_Propagation    avgt   10   85.074 ± 40.665  ms/op
[info] Benchmark results reported into /Users/gmt/dev/enso/enso/std-bits/benchmarks/./bench-report.xml
[success] Total time: 617 s (10:17), completed May 31, 2024, 4:38:10 PM
```

**Comment by GregoryTravis** [comment]
Results with @JaroslavTulach 's change. Improved in some cases, but still mysterious.

```
==> prop-flag-2-out <==
[info] Vector_Builder.Builder_New_Append_Batch_With_Warning_Propagation      avgt   10   28.875 ±  5.325  ms/op
[info] Vector_Builder.Builder_New_Append_Batch_Without_Warning_Propagation   avgt   10   16.300 ±  2.520  ms/op
[info] Vector_Builder.Builder_New_Append_One_With_Warning_Propagation        avgt   10  117.361 ± 51.105  ms/op
[info] Vector_Builder.Builder_New_Append_One_Without_Warning_Propagation     avgt   10  103.525 ± 35.385  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_With_Warning_Propagation     avgt   10   26.862 ±  5.193  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_Without_Warning_Propagation  avgt   10   14.910 ±  1.640  ms/op
[info] Vector_Builder.Vector_Build_Append_One_With_Warning_Propagation       avgt   10  194.968 ± 36.746  ms/op
[info] Vector_Builder.Vector_Build_Append_One_Without_Warning_Propagation    avgt   10   92.243 ± 56.763  ms/op
==> prop-flag-out <==
[info] Vector_Builder.Builder_New_Append_Batch_With_Warning_Propagation      avgt   10   25.885 ±  3.141  ms/op
[info] Vector_Builder.Builder_New_Append_Batch_Without_Warning_Propagation   avgt   10   20.868 ±  7.349  ms/op
[info] Vector_Builder.Builder_New_Append_One_With_Warning_Propagation        avgt   10  112.103 ± 39.995  ms/op
[info] Vector_Builder.Builder_New_Append_One_Without_Warning_Propagation     avgt   10  219.262 ± 70.423  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_With_Warning_Propagation     avgt   10   15.867 ±  1.989  ms/op
[info] Vector_Builder.Vector_Build_Append_Batch_Without_Warning_Propagation  avgt   10   18.376 ±  8.989  ms/op
[info] Vector_Builder.Vector_Build_Append_One_With_Warning_Propagation       avgt   10   90.453 ± 43.618  ms/op
[info] Vector_Builder.Vector_Build_Append_One_Without_Warning_Propagation    avgt   10   89.058 ± 51.438  ms/op
```

**Pull request merged**